includes/class-micropub-authorize.php

```diff
@@ -0,0 +1,218 @@
+			return isset( $array[ $key ] ) ? $array[ $key ] : $default;
+		}
+		return $default;
+	}
+
+	/**
+	 * Get the authorization header
+	 *
+	 * On certain systems and configurations, the Authorization header will be
+	 * stripped out by the server or PHP. Typically this is then used to
+	 * generate `PHP_AUTH_USER`/`PHP_AUTH_PASS` but not passed on. We use
+	 * `getallheaders` here to try and grab it out instead.
+	 *
+	 * @return string|null Authorization header if set, null otherwise
+	 */
+	public static function get_authorization_header() {
+		if ( ! empty( $_SERVER['HTTP_AUTHORIZATION'] ) ) {
+			return wp_unslash( $_SERVER['HTTP_AUTHORIZATION'] );
+		}
+		// When Apache speaks via FastCGI with PHP, then the authorization header is often available as REDIRECT_HTTP_AUTHORIZATION.
+		if ( ! empty( $_SERVER['REDIRECT_HTTP_AUTHORIZATION'] ) ) {
+			return wp_unslash( $_SERVER['REDIRECT_HTTP_AUTHORIZATION'] );
+		}
+		$headers = getallheaders();
+		// Check for the authorization header case-insensitively
+		foreach ( $headers as $key => $value ) {
+			if ( strtolower( $key ) === 'authorization' ) {
+				return $value;
+			}
+		}
+		return null;
+	}
+
+	/**
+	 * Validate the access token at the token endpoint.
+	 *
+	 * https://indieauth.spec.indieweb.org/#access-token-verification
+	 */
+	public static function determine_current_user( $user_id ) {
+		$cls = get_called_class();
+		// Do not try to find a user if one has already been found
+		if ( ! empty( $user_id ) ) {
+			return $user_id;
+		}
+
+		// Only try to authenticate if this is the Micropub endpoint
+		if ( ! empty( get_query_var( 'micropub' ) ) ) {
+			return $user_id;
+		}
+
+		// find the access token
+		$auth  = static::get_authorization_header();
+		$token = self::get( $_POST, 'access_token' );
+		if ( ! $auth && ! $token ) {
+			static::authorize_error( 401, 'missing access token' );
+		}
+
+		$resp = wp_remote_get(
+			get_option( 'indieauth_token_endpoint', MICROPUB_TOKEN_ENDPOINT ), array(
+				'headers' => array(
+					'Accept'        => 'application/json',
+					'Authorization' => $auth ?: 'Bearer ' . $token,
+				),
+			)
+		);
+		if ( is_wp_error( $resp ) ) {
+			return $resp;
+		}
+
+		$code           = wp_remote_retrieve_response_code( $resp );
+		$body           = wp_remote_retrieve_body( $resp );
+		$params         = json_decode( $body, true );
+		static::$scopes = explode( ' ', $params['scope'] );
+
+		if ( (int) ( $code / 100 ) !== 2 ) {
+			return static::authorize_error(
+				$code, 'invalid access token: ' . $body
+			);
+		} elseif ( empty( static::$scopes ) ) {
+			return static::authorize_error(
+				401, 'access token is missing scope'
+			);
+		}
+
+		$me                             = untrailingslashit( $params['me'] );
+		static::$micropub_auth_response = $params;
+
+		// look for a user with the same url as the token's `me` value.
+		$user = static::user_url( $me );
+
+		add_filter( 'indieauth_scopes', array( $cls, 'indieauth_scopes' ), 11 );
```

Comment:
it seems a bit surprising that these happen here, lazily during a request, instead of in `init`. mind moving them there?

Reply:
ping!